src/Parsers/ExpressionListParsers.cpp

```diff
@@ -599,158 +609,1974 @@ bool ParserLambdaExpression::parseImpl(Pos & pos, ASTPtr & node, Expected & expe
+        result.push_back(std::move(op));
+    }
+
+    virtual bool getResult(ASTPtr & op)
+    {
+        if (result.size() == 1)
+        {
+            op = std::move(result[0]);
+            return true;
+        }
+
         return false;
+    }
 
-    if (!*it)
-        node = elem;
-    else
+    virtual bool parse(IParser::Pos & /*pos*/, Expected & /*expected*/, Action & /*action*/)
     {
-        for (size_t i = 0; i < count; ++i)
+        return true;
+    }
+
+    bool isFinished() const
+    {
+        return state == -1;
+    }
+
+    int previousPriority() const
+    {
+        if (operators.empty())
+            return 0;
+
+        return operators.back().priority;
+    }
+
+    OperatorType previousType() const
+    {
+        if (operators.empty())
+            return OperatorType::None;
+
+        return operators.back().type;
+    }
+
+    int empty() const
+    {
+        return operators.empty() && operands.empty();
+    }
+
+    bool lastNOperands(ASTs & asts, size_t n)
+    {
+        if (n > operands.size())
+            return false;
+
+        auto start = operands.begin() + operands.size() - n;
+        asts.insert(asts.end(), std::make_move_iterator(start), std::make_move_iterator(operands.end()));
+        operands.erase(start, operands.end());
+
+        return true;
+    }
+
+    bool wrapLayer(bool push_to_result = true)
+    {
+        Operator cur_op;
+        while (popOperator(cur_op))
         {
-            /// the function corresponding to the operator
-            auto function = std::make_shared<ASTFunction>();
+            ASTPtr func;
 
-            /// function arguments
-            auto exp_list = std::make_shared<ASTExpressionList>();
+            // Special case of ternary operator
+            if (cur_op.type == OperatorType::StartIf)
+                return false;
 
-            function->name = it[1];
-            function->arguments = exp_list;
-            function->children.push_back(exp_list);
+            if (cur_op.type == OperatorType::FinishIf)
+            {
+                Operator tmp;
+                if (!popOperator(tmp) || tmp.type != OperatorType::StartIf)
+                    return false;
+            }
 
-            if (node)
-                exp_list->children.push_back(node);
+            // Special case of a BETWEEN b AND c operator
+            if (cur_op.type == OperatorType::StartBetween || cur_op.type == OperatorType::StartNotBetween)
+                return false;
+
+            if (cur_op.type == OperatorType::FinishBetween)
+            {
+                Operator tmp_op;
+                if (!popOperator(tmp_op))
+                    return false;
+
+                if (tmp_op.type != OperatorType::StartBetween && tmp_op.type != OperatorType::StartNotBetween)
+                    return false;
+
+                bool negative = tmp_op.type == OperatorType::StartNotBetween;
+
+                ASTs arguments;
+                if (!lastNOperands(arguments, 3))
```

Comment:
When we return `false` here it means that the original query was not correct. Therefore a change in the state does not really matter, since parsing has failed.